packages/contracts/contracts/optimistic-ethereum/ovm/SafetyChecker.sol

```diff
@@ -63,136 +35,102 @@ contract SafetyChecker is ContractResolver {
-                            gasOp > 0x8f || // DUPs are 0x80...0x8f
-                            addressOp != 0x73 || // address must be set with a PUSH20
-                            valueOp != 0x60 // value must be set with a PUSH1
-                        ) {
-                            return false;
-                        } else {
-                            uint256 pushedBytes;
-                            // gas is set with a PUSH##
-                            if (gasOp >= 0x60 && gasOp <= 0x7f) {
-                                pushedBytes = gasOp - 0x5f;
-                            }
+                let mpc := byte(0, mload(add(skip, byte(0, tmp))))
+                mpc := add(mpc, byte(0, mload(add(skip, byte(mpc, tmp)))))
+                mpc := add(mpc, byte(0, mload(add(skip, byte(mpc, tmp)))))
+                mpc := add(mpc, byte(0, mload(add(skip, byte(mpc, tmp)))))
+                mpc := add(mpc, byte(0, mload(add(skip, byte(mpc, tmp)))))
+                mpc := add(mpc, byte(0, mload(add(skip, byte(mpc, tmp)))))
+                _pc := add(_pc, mpc)
 
-                            // 23 is from 1 + PUSH20 + 20 bytes of address + PUSH or DUP gas
-                            byte callValue = _bytecode[pc - (23 + pushedBytes)];
-
-                            // 21 is from 1 + 19 bytes of address + PUSH or DUP gas
-                            address callAddress = toAddress(_bytecode, (pc - (21 + pushedBytes)));
+                op := byte(mpc, tmp)
+            }
 
-                            // CALL is made to the execution manager with msg.value of 0 ETH
-                            if (callAddress != address(resolveExecutionManager()) || callValue != 0 ) {
-                                return false;
-                            }
+            // + push opcodes
+            // + stop opcodes [STOP(0x00),JUMP(0x56),RETURN(0xf3),REVERT(0xfd),INVALID(0xfe)]
+            // + caller opcode CALLER(0x33)
+            // + blacklisted opcodes
+            uint256 opBit = 1 << op;
+            if (opBit & _opcodeProcMask == 0) {
+                if (opBit & _opcodePushMask == 0) {
+                    // subsequent bytes are not opcodes. Skip them.
+                    _pc += (op - 0x5e);
+                    // all pushes are valid opcodes
+                    continue;
+                } else if (opBit & _opcodeStopMask == 0) {
+                    // STOP or JUMP or RETURN or REVERT or INVALID (see safety checker docs in wiki for more info)
+                    // We are now inside unreachable code until we hit a JUMPDEST!
+                    do {
+                        _pc++;
+                        assembly {
+                            op := byte(0, mload(_pc))
                         }
+                        // encountered a JUMPDEST
+                        if (op == 0x5b) break;
+                        // skip PUSHed bytes
+                        if ((1 << op) & _opcodePushMask == 0) _pc += (op - 0x5f);
```

Comment:
Interesting... so even in unreachable code, we have to honor `PUSH##`? It's impossible for unreachable code to have any of the `PUSH##` opcodes without the correct number of bytes following it?

Reply:
If there's a `JUMPDEST` inside of `PUSH`ed bytes, then it's still unreachable code. If the bytecode ends in unreachable code and ends with, say a `PUSH32` with less than 32 bytes following it, then it's still safe. This will be handled by the `while (pc < codeLength);`